corehq/apps/app_manager/views/apps.py

```diff
@@ -416,25 +417,23 @@ def _inner(request, link_domain, data, master_domain=domain):
     return login_and_domain_required(_inner)(request, form.cleaned_data['domain'], form.cleaned_data)
 
 
-def _create_linked_app(request, master_app, link_domain, link_app_name, master_build_id=None):
-    master_domain = master_app.domain
-    if master_build_id:
-        master_version = get_app(master_domain, master_build_id)
-    else:
-        master_version = get_latest_released_app_version(master_domain, master_app._id)
-    if not master_version:
+def _create_linked_app(request, master_build, link_domain, link_app_name):
+    master_domain = master_build.domain
+    if not master_build:
```

Comment:
Not sure what the correct approach is, but this needs some rethinking - if `master_build` is `None`, `master_build.domain` will fail above this, and `master_build.master_id` will fail below.

Reply:
Clearly I'm having trouble getting my head all the way around `copy_app`. I took a fresh look at it and refactored in https://github.com/dimagi/commcare-hq/pull/25003/commits/ad3fb0a32d62a62cec26eb949221f8aeba2a307e with two goals: 1) move as much logic as possible inside `_inner`, rather than having some logic before it and then some inside and 2) make the function take two separate paths (normal vs linked app) that don't share much logic.